### DISCUSSION TITLE
Is it possible to enhance the static site generator?

### DISCUSSION THREAD

**Discussion by hysonlee** [datenstrom/community#125] [Ask a question] [answered]
Hi,
After reading [an article](http://www.smashingmagazine.com/2015/11/modern-static-website-generators-next-big-thing/) about the static website and based on my previous experience in Yellow static builder, I have an opinion: if Yellow builder works fine as a full static generator that there is no dependency on the `htaccess` file, hosting a static site that will be more convenient, and easer.

**Comment by markseuffert** [comment]
Can you explain the issues with Yellow as static site generator? Something not working on your server?

**Comment by hysonlee** [comment]
Not a issue, I just want the static site works without a `htaccess` (rewrite),  namely output the page file as `page.html`.

**Comment by markseuffert** [comment]
Datenstrom Yellow is a flat file CMS and a static site generator.  Let's walk trough the current way of working: 

One of the design goals was to make websites and it shouldn't matter if you serve them dynamically or statically. Basically, the website you see in the CMS is a website that you can also generate statically, upload to a webserver and use without CMS. Makes things super fast and safe. There are a few limitations though, for example a static site generator can't provide online editing or text search. Keep in mind that Yellow can operate in both ways, dynamically and statically. I don't know about many web publishing systems that work in both ways. This might explain a few things that are done differently.

If Yellow is used as a **flat file CMS** then web pages will be generated on-the-fly. The web server will redirect (many) requests to the CMS, which converts files and sends them back to the browser. On the Apache webserver the `.htaccess` is used to rewrite requests, see rules marked "Yellow dynamic pages". A missing file will be handled by the CMS, you can define how this will looks like.

If Yellow is used as a **static site generator** then web pages will be served as they are. The web server will send files directly from the file system back to the browser. On the Apache webserver the `.htaccess` is used to support custom 404 page, see rules marked "Yellow static pages". A missing file will be handled directly by the web server, which can use the page generated by Yellow.

```
<IfModule mod_rewrite.c>
RewriteEngine on

# Yellow dynamic pages

RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule ^ yellow.php [L]
DirectoryIndex index.html yellow.php
RewriteRule ^(cache|content|system)/ error [L]

# Yellow static pages

RewriteCond %{REQUEST_FILENAME} -f
RewriteRule ^[^\.]+$ - [T=text/html,L]
ErrorDocument 404 /error.html
</IfModule>
```

Hope this explains what is happening. Yellow doesn't have to work this way, suggestions are welcome and everything that makes things easier for people is welcome.

**Comment by hysonlee** [comment]
Of course I understand the basic principle of YellowCMS how it woks, I'm afraid you have misunderstood my point. After the day's exploration and debugging , I think I know how to achieve what i mentioned above(a little modify to the `commandline` script).  
Please forget it and go ahead to follow the development blueprint of Yellow.

Appreciate your patient explanation again Mark. Have a nice day. ;)

**Comment by markseuffert** [comment]
Sorry, if I have misunderstood you. Please let me know how I can help. :)

**Comment by markseuffert** [comment, marked as answer]
A summary of what we have discussed via email:
- If an auto-generated htaccess file is not flexible and sometimes not supported, then there's no need to include it in static output. 
- Windows doesn’t like the colon character in the URL, therefor fail fast with error 500 and in the future we could support Windows file system to make Windows user happy. 
- HTML pages without html-suffix work on all tested web servers out-of-the-box, with an exception of Microsoft's IIS which is considered an unsupported web server for now. 
- Pages with footnotes create unnecessary files, where file name contains "#fn", something to be fixed.

Let's see if users report any problems. Please comment if something is wrong or missing.

**Comment by wunderfeyd** [comment]
Which character do you suggest for windows file systems as separator? Maybe one could use multiple character too? Okay, this depends on the general encoding rules of the file names and the normal http rules. What about sequences like a double hyphen or anything similiar?

**Comment by markseuffert** [comment]
My suggestion is to not support Windows file system for the moment. Instead fail fast with a clear error message: Static website not supported on Windows file system. In the future we could use a different separator, one alternative could be using `=` or `;` instead of a colon.

**Comment by markseuffert** [comment]
I updated server configuration and commandline based on feedback from @hysonlee.

**Comment by hysonlee** [comment]
I have tried with the latest Yellow repository, both the `build` and `clean` command work fine, there's no more error except the `contact` page as we know. Also the issue "Pages with footnotes create unnecessary files" closed. Here is my complete record:

```
hyson@hyson-mint /mnt/shared/htdocs $ php yellow.php version
Yellow 0.6.1
YellowBlog 0.5.7
YellowBreadcrumbs 0.5.2
YellowCommandline 0.6.2
YellowContact 0.5.2
YellowCore 0.6.1
YellowDisqus 0.6.1
YellowDraft 0.6.1
YellowFeed 0.6.1
YellowFontawesome 0.5.3
YellowFotorama 0.5.2
YellowImage 0.5.1
Yellowiqiniu 0.5.1
YellowMarkdown 0.6.1
YellowSitemap 0.5.4
YellowToc 0.5.2
YellowWebinterface 0.6.1
YellowWiki 0.5.7
YellowWymusic 0.5.1
YellowYoutube 0.6.1
ERROR checking updates at https://github.com/datenstrom/yellow-extensions: Plugin 'commandline' requires cURL library!
hyson@hyson-mint /mnt/shared/htdocs $ php yellow.php build /home/hyson/htdoc 
ERROR building content location '/contact/', HTTP/1.1 500 Server error: Static website not supported!
Yellow build: 152 content, 20 media, 6 system, 1 error, status 500
hyson@hyson-mint /mnt/shared/htdocs $ php yellow.php clean /home/hyson/htdoc 
Yellow clean: Static files cleaned
hyson@hyson-mint /mnt/shared/htdocs $ php yellow.php build /home/hyson/htdoc 
ERROR building content location '/contact/', HTTP/1.1 500 Server error: Static website not supported!
Yellow build: 152 content, 20 media, 6 system, 1 error, status 500
hyson@hyson-mint /mnt/shared/htdocs $ 
```

To define a 404 page for Yellow static site , just create a `.htaccess` file and add a line as `ErrorDocument 404 /404.html`. It works for me.
Above, thanks Mark!

**Comment by markseuffert** [comment]
Good to hear that it works, thanks for your help Hyson!
